File: 206_reverse_linked_list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
    if (!head || !head.next){
        return head;
    }

    const storage = [];
    let currentNode = head;
    while (currentNode){
        storage.push(currentNode);
        currentNode = currentNode.next;
    }
    let output;
    while (storage.length > 0){
        const poppedNode = storage.pop();
        if (!output){
            output = poppedNode;
            currentNode = output;
        } else {
            currentNode.next = poppedNode;
            currentNode = poppedNode;
        }
    }
    currentNode.next = null;
    return output;
    //Time Complexity O(n), Space Complexity O(n)
};

File: 206_reverse_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
        
        storage = []
        currentNode = head
        while currentNode is not None:
            storage.append(currentNode)
            currentNode = currentNode.next
        output = None
        while len(storage) > 0:
            poppedNode = storage.pop()
            if output is None:
                output = poppedNode
                currentNode = output
            else:
                currentNode.next = poppedNode
                currentNode = poppedNode
        currentNode.next = None
        return output